File: c/cpu scheduling/sjf.c
# include <stdio.h>
# include <stdlib.h>
# include <time.h>

typedef struct {
    int pno,at,bt,wt,tt,flag;
}process;

int randint(int lower,int upper){
    return rand()%(upper-lower+1)+lower;
}
void disp(process p[],int n){
    int i;
    printf("Pno\tarrival\tBurst\tWait\tTurn Around\tflag");
    for(i=0;i<n;i++)
        printf("\n%d\t%d\t%d\t%d\t%d\t\t%d\n",p[i].pno,p[i].at,p[i].bt,p[i].wt,p[i].tt,p[i].flag);
}
void calc_wt(process p[],int n){
    int i,ct,finish_count,minbt,next_process, at_flag = 0;
    float ttt,twt = 0 ;
    ct = p[0].at + p[0].bt;
    p[0].tt = p[0].bt;
    ttt = p[0].bt;
    p[0].flag = 1; //First Process Complete
    finish_count = 1;
    printf("\npno = %d ct = %d arrival = %d waiting = %d ",p[0].pno,ct,p[0].at,p[0].wt);
    while(finish_count < n){
        at_flag = 0;
        minbt = 99999;
        for(i=1;i<n;i++){
            if(p[i].flag == 1)
                continue;
            else if(p[i].at < ct){
                at_flag = 1;
                if(p[i].bt < minbt){
                    minbt = p[i].bt;
                    next_process = i;
                }
                    
            }
            else if(at_flag == 0){
                next_process = i;
                break;
            }        
        }

    if(at_flag == 1){
        p[next_process].wt = ct - p[next_process].at;
        ct += p[next_process].bt;
    }
    else
        ct = p[next_process].at + p[next_process].bt;
    printf("\npno = %d ct = %d arrival = %d waiting = %d ",p[next_process].pno,ct,p[next_process].at,p[next_process].wt);
    p[next_process].flag = 1;
    finish_count++;
    p[next_process].tt = p[next_process].wt + p[next_process].bt;    
    twt += p[next_process].wt;    
    ttt += ( p[next_process].wt + p[next_process].bt );    
    }
    printf("\nAverage waiting time = %f",twt/n);
    printf("\nAverage Turn around time time = %f\n",ttt/n);

}

int compare(const void *p,const void *q){
    int at1 = ((process *)p)->at;
    int at2 = ((process *)q)->at;
    int bt1 = ((process *)p)->bt;
    int bt2 = ((process *)q)->bt;
    if(at1 == at2)
        return bt1-bt2;
    else
        return at1-at2;        
}

int main(){
    int n,i;
    srand(time(NULL));
    n = randint(1,10);
    process *p = malloc(n*sizeof(process));
    for(i=0;i<n;i++){
        p[i].pno = i+1;
        p[i].bt  = randint(1,25);
        p[i].at  = randint(0,25);
        p[i].flag = 0;
        p[i].wt  = 0;
        p[i].tt  = 0;
    }    
    disp(p,n);  
    qsort(p,n,sizeof(process),compare);  
    calc_wt(p,n);
    disp(p,n);
    return 0;
}

File: python/cpu scheduling/sjf.py
from random import randint
from operator import add

# utility function


def disp(process):
    print('pno\tArival\tBurst')
    print('*'*25)
    for i in process:
        for j in i:
            print(j, end='\t')
        print('\n')


n = int(input('Enter number of process :'))
# creating process using randint
process_list = [[i, randint(1, 20), randint(1, 20)] for i in range(n)]
print('\nProcess list')
disp(process_list)
# sort with arival time
process_list.sort(key=lambda x: (x[1], x[2]))

sjf = []
wt = [0]
bt = [process[2] for process in process_list]
current_time = process_list[0][1]+process_list[0][2]
sjf.append(process_list.pop(0))  # first  arrived process

for _ in range(len(process_list)):
    temp = [p for p in process_list if p[1] <= current_time]
    if temp:
        next_process = min(temp, key=lambda x: x[2])
        process_list.remove(next_process)
    else:
        next_process = process_list.pop(0)

    # incrementing current
    if next_process[1] <= current_time:
        wt.append(current_time-next_process[1])
        current_time += next_process[2]
    else:
        wt.append(0)
        current_time = (next_process[1]+next_process[2])
        
    sjf.append(next_process)

print('SJF')
disp(sjf)
tt = list(map(add, wt, bt))
print(f'Average waiting time = {sum(wt)/n}')
print(f'Average Turn around time time = {sum(tt)/n}')
